### PR TITLE
[BREAKING] Package rewrite for a 1.0 release

### PR DESCRIPTION
**Pull request by quinnj**
This is a proposal to overhaul the package internals, providing mostly the same high-level interfaces (`JSON.parse`, `JSON.json`, etc) while also adding a number of enhancements (lazy parsing support, typed materialization similar to JSON3).

My goal is to deprecate JSON3.jl in favor of the simpler, faster, and overall better functionality as implemented in this PR.

I've reviewed all existing open issues/PRs to both [this package](https://github.com/JuliaIO/JSON.jl/issues?q=is%3Aissue%20state%3Aopen%20label%3A%22Proposed%20fix%20in%201.0%20release%22) and [JSON3.jl](https://github.com/quinnj/JSON3.jl/issues?q=is%3Aissue%20state%3Aopen%20label%3A%22Fixed%20in%20JSON.jl%201.0%22), marking all that will be fixed or resolved with the functionality as proposed (over 70 issues resolved in total!).

I've tried to provide extensive new documentation (in the new docs/ directory) on all proposed functionality, including a migrate.md file that provides a detailed migration guide for both JSON.jl pre-1.0 users and JSON3.jl users.

My hope and aim is that the net upgrade for the vast majority of JSON pre-1.0 users will simply be adding `JSON = "1"` in their compat files. I've tried to support as much as possible in what I could tell was part of the JSON.jl public API. A few things were deliberately left out as my impression is they were either part of an ancient API that can be better served by more modern Julia mechanisms, or just seem not very useful. If there are things we discover feel too breaking, I'm definitely open to try and find ways to add backwards compatibility.

Another part of my commitment in this proposal, if/when merged and released, is to take time across the ecosystem to help upgrade packages. I've done this for other packages (CSV.jl, DataFrames.jl, HTTP.jl, etc.) with big releases, and I feel it helps with a new release "momentum" to get a bunch of key packages upgraded on the new release.

This proposal does add another dependency on the newish StructUtils.jl package and I'll try to provide a little context/justification. I've actually been working on the core refactor of this code and StructUtils.jl for around 2 years. The ideas and redesign came from architectural pains/complexities in JSON3.jl and StructTypes.jl, and a desire to find an internal framework that was at least as powerful/flexible as what ST + J3 provided, but in a vastly simpler way (the JSON3.jl code kept growing in complexity and became hard to work with/maintain/improve). I believe the design of StructUtils.jl to be much less invasive and natural when working with Julia structs, while the power of the core functional parsing methods provide a clean integration for handling 1) default materialization from JSON, 2) typed materialization, and 3) simple serialization from structs and a variety of other types. What's more, is the StructUtils.jl machinery generalizes easily beyond just JSON, and I already have other unreleased packages that use it for database model interactions, struct diffing, and a number of other use-cases.

As for review, I recognize that huge PRs like this are very hard to break down and see the exact net changes. In this case, the entire package is pretty much rewritten, so I'd recommend checking out the PR locally, testing out code, and reviewing the entire package/files in that way (i.e. trying to look at any kind of "diff" on github isn't probably that useful).

I'd also like to propose a public "community review" time where I would host a google meet call that would be open to anyone interested and I would take time to go over package internals/architecture and help answer questions/concerns in person. If folks are interested, please ping me in the public Julia slack or comment here and I'll arrange a time that works for those interested.

### CODE REVIEW

File: docs/src/migrate.md [hunk 1]
```diff
@@ -0,0 +1,236 @@
...
+    - Can now pass `JSON.json(x; pretty=true)` or `JSON.json(x; pretty=4)` to control pretty printing
+    - Can pass filename as first argument to write JSON directly to a file `JSON.json(file, x)` the file name is returned
+    - Can pass any `IO` as 1st argument to write JSON to it: `JSON.json(io, x)`
+    - Circular reference tracking is fixed/improved (previously peer references were written as null)
+    - Explicit keyword arguments to control a number of serialization features, including:
+      - `omit_null::Bool` whether `nothing`/`missing` Julia values should be skipped when serializing
+      - `omit_empty::Bool` whether empty Julia collection values should be skipped when serializing
+      - `allownan::Bool` similar to the parsing keyword argument to allow/disallow writing of invalid JSON values `NaN`, `-Inf`, and `Inf`
+      - `ninf::String` the string to write if `allownan=true` and serializing `-Inf`
+      - `inf::String` the string to write if `allownan=true` and serializing `Inf`
```
Comment (oscardssmith) [inline docs/src/migrate.md:25]: possibly `inf_string`? Also do we need `ninf` or can it just print as a `-` before `inf`?

File: docs/src/migrate.md [hunk 2]
```diff
@@ -0,0 +1,236 @@
...
+    - Can take an integer 2nd argument to induce "pretty printing" of the JSON output
+  - What changed:
+    - Can now pass `JSON.json(x; pretty=true)` or `JSON.json(x; pretty=4)` to control pretty printing
+    - Can pass filename as first argument to write JSON directly to a file `JSON.json(file, x)` the file name is returned
+    - Can pass any `IO` as 1st argument to write JSON to it: `JSON.json(io, x)`
+    - Circular reference tracking is fixed/improved (previously peer references were written as null)
+    - Explicit keyword arguments to control a number of serialization features, including:
+      - `omit_null::Bool` whether `nothing`/`missing` Julia values should be skipped when serializing
+      - `omit_empty::Bool` whether empty Julia collection values should be skipped when serializing
+      - `allownan::Bool` similar to the parsing keyword argument to allow/disallow writing of invalid JSON values `NaN`, `-Inf`, and `Inf`
```
Comment (oscardssmith) [inline docs/src/migrate.md:23]: also, should this be `allownonfinite`?

Comment (quinnj) [reply docs/src/migrate.md:23]: JSON3.jl had `allow_inf`, but JSON.jl already had `allownan`, so I opted to keep that for backwards compat.

**Comment by ericphanson** [comment]
Tiny detail but these methods don’t close the file handle: https://github.com/JuliaIO/JSON.jl/blob/c9d0b1b2e26a7cd8427fd83d7332b95826f8444c/src/parse.jl#L172

**Comment by quinnj** [comment]
Interesting situation w/ the Documentation job failure here: Documenter depends on JSON, but doesn't have compat w/ 1.0, so it can't install. Would we need to have Documenter.jl pre-emptively support 1.0? Other solutions?

File: docs/src/migrate.md (hunk 2 above)
Comment (tecosaur) [reply docs/src/migrate.md:23]: Could it be worth checking how many packages depend on the current JSON.jl to estimate how breaking any "breaking changes" might be in practice?

**Comment by KristofferC** [comment]
JSON ends up being a dependency in most environments. Is there any significant precompile / load times changes with this PR?

**Comment by quinnj** [comment]
> JSON ends up being a dependency in most environments. Is there any significant precompile / load times changes with this PR?

```
# 1.0 precompile:
julia> @time using JSON
Precompiling JSON...
  1 dependency successfully precompiled in 3 seconds. 11 already precompiled.
  3.074514 seconds (243.41 k allocations: 17.114 MiB, 1.44% compilation time)


# 1.0 load: 
julia> @time using JSON
  0.067531 seconds (119.51 k allocations: 7.854 MiB)


# Pre-1.0 load:
julia> @time using JSON
  0.181937 seconds (54.85 k allocations: 3.523 MiB)



# pre-1.0 precompile:
julia> @time using JSONold
Precompiling JSONold...
  1 dependency successfully precompiled in 1 seconds. 9 already precompiled.
  1.039715 seconds (95.82 k allocations: 7.128 MiB, 1.91% compilation time)
```

**Comment by quinnj** [comment]
> Tiny detail but these methods don’t close the file handle:
> 
> https://github.com/JuliaIO/JSON.jl/blob/c9d0b1b2e26a7cd8427fd83d7332b95826f8444c/src/parse.jl#L172

Just pushed a fix; thanks for mentioning.

**Comment by KristofferC** [comment]
Having to uncompress all the test files for all package updates feels like it could be slightly slow. In TOML (which has a similar test suite) I get them during testing from an archive (https://github.com/JuliaLang/TOML.jl/blob/44aab3c023323587680ab8d8c7ef478bf78c4c0c/test/utils/utils.jl#L8) but I guess an archive could also be checked in. Alternatively, all the content of the files could be put into one file with a sentinel "file separator" and get constructed on the fly during testing.

**Comment by quinnj** [comment]
> Having to uncompress all the test files for all package updates feels like it could be slightly slow (at least on Windows). In TOML (which has a similar test suite) I get them during testing from an archive (https://github.com/JuliaLang/TOML.jl/blob/44aab3c023323587680ab8d8c7ef478bf78c4c0c/test/utils/utils.jl#L8) but I guess an archive could also be checked in. Alternatively, all the content of the files could be put into one file with a sentinel "file separator" and get constructed on the fly during testing.

Great callout; I refactored so that we have tar files of both `jsonchecker` and `JSONTestSuite` test directories that get untarred when tests are run. I think that it's setup fine? Not sure if there are other considerations or if it's bad practice to untar in the test directory?

**Comment by tecosaur** [comment]
> that get untarred when tests are run

What about just getting each file from the tarball directly (no on-filesystem untaring) using Tar.jl?

**Comment by quinnj** [comment]
> > that get untarred when tests are run
> 
> What about just getting each file from the tarball directly (no on-filesystem untaring) using Tar.jl?

Ah, can Tar.jl do that? I was going off the README docs and I didnt' see anything about that, but that would indeed by nice.

**Comment by quinnj** [comment]
> > that get untarred when tests are run
> 
> What about just getting each file from the tarball directly (no on-filesystem untaring) using Tar.jl?

Ok, I looked into `Tar.extract` with the predicate, but it seems pretty inefficient (checking each file in tarball for each extraction) and you still have to "extract" the single file to a directory. If there was a way to extract to an IOBuffer, I feel like that would make sense, but otherwise, I'm thinking of keeping it as-is.

**Comment by tecosaur** [comment]
You can create a `Dict{String, Vector{UInt8}}` for each file like so: https://github.com/tecosaur/DataToolkit.jl/blob/main/Common/ext/TarExt.jl#L22-L34

**Comment by quinnj** [comment]
> You can create a `Dict{String, Vector{UInt8}}` for each file like so: https://github.com/tecosaur/DataToolkit.jl/blob/main/Common/ext/TarExt.jl#L22-L34

Ah! Thanks for the reference; that seems indeed to work well. I've added a little utility like your extension. It'd be nice if Tar.jl itself provided this as a function.

**Comment by KristofferC** [comment]
Extracting to a temp folder or a scratch space or something would at least be better than the test folder, better leave the package sources unmodified.

**Comment by quinnj** [comment]
A goole meet video conference has been setup for tomorrow (Friday), May 16 at 11:30 AM EST for those interested in discussing the proposed JSON.jl 1.0 release. I'll give a brief summary of the proposed changes, but mostly spend time answering questions and getting feedback. Please join if you can! The more the merrier! If there's enough interest for folks who can't make this time, please reach out and we can schedule another at a time that works.

File: src/lazy.jl [hunk 3]
```diff
@@ -0,0 +1,722 @@
...
+@label invalid
+    invalid(error, buf, pos, "array")
+end
+
+# temporary string type to enable deferrment of string allocation in certain cases (like navigating a lazy structure)
+struct PtrString
+    ptr::Ptr{UInt8}
+    len::Int
+    escaped::Bool
+end
```
Comment (nhz2) [inline src/lazy.jl:410]: Why not use a view instead of a ptr and len here?

File: src/lazy.jl [hunk 4]
```diff
@@ -0,0 +1,722 @@
...
+
+if VERSION < v"1.11"
+    mem(n) = Vector{UInt8}(undef, n)
+    _tostr(m::Vector{UInt8}, slen) = ccall(:jl_array_to_string, Ref{String}, (Any,), resize!(m, slen))
```
Comment (nhz2) [inline src/lazy.jl:414]: These are not public functions, and also I'm pretty sure that in this case they won't have any performance advantage over using `unsafe_string`.

File: src/lazy.jl [hunk 5]
```diff
@@ -0,0 +1,722 @@
...
+function Base.convert(::Type{String}, x::PtrString)
+    if x.escaped
+        m = mem(x.len)
+        slen = GC.@preserve m unsafe_unescape_to_buffer(x.ptr, x.len, pointer(m))
+        return _tostr(m, slen)
+    end
+    return unsafe_string(x.ptr, x.len)
+end
+
+Base.convert(::Type{Symbol}, x::PtrString) = ccall(:jl_symbol_n, Ref{Symbol}, (Ptr{UInt8}, Int), x.ptr, x.len)
```
Comment (nhz2) [inline src/lazy.jl:429]: `jl_symbol_n` is not public.

File: src/utils.jl [hunk 6]
```diff
@@ -0,0 +1,249 @@
...
+        NULL => "NULL",
+        NUMBER => "NUMBER",
+    )
+    Base.show(io::IO, x::T) = Base.print(io, "JSONTypes.", names[x])
+end
+
+# windows doesn't like mmaped buffers hanging around and will error if you try to modify
+# so we call finalize eagerly to ensure the buffer is unmapped
+# ref: https://github.com/JuliaIO/JSON.jl/issues/112
+checkfinalize!(buf) = Sys.iswindows() && (VERSION ≥ v"1.11" ? finalize(buf.ref.mem) : finalize(buf))
```
Comment (nhz2) [inline src/utils.jl:44]: I think the `ref` field is not public.

File: src/write.jl [hunk 7]
```diff
@@ -0,0 +1,621 @@
+struct JSONWriteStyle <: JSONStyle end
+
+sizeguess(::Nothing) = 4
+sizeguess(x::Bool) = 5
+sizeguess(x::Integer) = 20
+sizeguess(x::AbstractFloat) = 20
+sizeguess(x::Base.IEEEFloat) = Base.Ryu.neededdigits(typeof(x))
```
Comment (nhz2) [inline src/write.jl:7]: From what I can tell, both `Base.IEEEFloat` and `Base.Ryu.neededdigits` are not public.

File: src/write.jl [hunk 8]
```diff
@@ -0,0 +1,621 @@
...
+with double quotes around it.
+
+An example overload would look something like:
+```julia
+JSON.tostring(x::MyDecimal) = string(x)
+```
+"""
+tostring(x) = string(Float64(x))
+
+_split_sign(x) = Base.split_sign(x)
```
Comment (nhz2) [inline src/write.jl:540]: `Base.split_sign` is not public.

File: src/write.jl [hunk 9]
```diff
@@ -0,0 +1,621 @@
...
+_split_sign(x::BigInt) = (abs(x), x < 0)
+
+@noinline infcheck(x, allownan) = isfinite(x) || allownan || throw(ArgumentError("$x not allowed to be written in JSON spec; pass `allownan=true` to allow anyway"))
+
+function _number(buf, pos, x::Number, opts::WriteOptions)
```
Comment (nhz2) [inline src/write.jl:545]: From what I can tell, all the `Base.Ryu` stuff in this function is not public.

File: src/parse.jl [hunk 10]
```diff
@@ -0,0 +1,361 @@
...
+parsefile!(file, x::T; jsonlines::Union{Bool, Nothing}=nothing, kw...) where {T} = open(io -> parse!(io, x; jsonlines=(jsonlines === nothing ? isjsonl(file) : jsonlines), kw...), file)
+
+parse(io::Union{IO, Base.AbstractCmd}, ::Type{T}=Any; kw...) where {T} = parse(Base.read(io), T; kw...)
+
+function parse(io::IOStream, ::Type{T}=Any; kw...) where {T}
+    buf = Mmap.mmap(io)
+    res = parse(buf, T; kw...)
+    checkfinalize!(buf)
+    return res
+end
```
Comment (nhz2) [inline src/parse.jl:183]: I don't think `mmap` can be reliably supported here, because `checkfinalize!(buf)` currently needs to access private fields of `buf`.

**Review by nhz2 — CHANGES_REQUESTED** [review]

File: src/lazy.jl [hunk 11]
```diff
@@ -0,0 +1,722 @@
...
+    allownan::Bool = false
+    ninf::String = "-Infinity"
+    inf::String = "Infinity"
+    nan::String = "NaN"
+    jsonlines::Bool = false
+end
+
+lazy(io::Union{IO, Base.AbstractCmd}; kw...) = lazy(Base.read(io); kw...)
+# we don't call checkfinalize! here because LazyValue holds a ref to buf for it's lifetime
+lazy(io::IOStream; kw...) = lazy(Mmap.mmap(io); kw...)
```
[the remark this answers is not in the document]
Comment (quinnj) [reply src/lazy.jl:81]: I've enhanced the docs a bit for `JSON.lazy` to make sure it's even more explicit what's happening (it was already mostly spelled out).

File: src/lazy.jl [hunk 12]
```diff
@@ -0,0 +1,722 @@
...
+@doc (@doc lazy) lazyfile
+
+function lazy(buf::Union{AbstractVector{UInt8}, AbstractString}; kw...)
+    if !applicable(pointer, buf, 1)
+        if buf isa AbstractString
+            buf = String(buf)
+        else
+            buf = Vector{UInt8}(buf)
+        end
+    end
```
[the remark this answers is not in the document]
Comment (quinnj) [reply src/lazy.jl:94]: Great call on the strides check. I've added that to `JSON.lazy`. I think in general, the JSON.jl and JSON3.jl libraries have taken the approach of trying to what the user would probably have to to themselves if their input isn't quite right (calling `read(io)` or `String(buf)` here), so I'm inclined to leave the behavior vs. throwing an error. I think it would leave to a more disruptive breaking release if JSON.jl now became much more ticky-tacky about what's allowed vs. not. There's also the ability to still customize the behavior for certain strings/vectors if there's a particular case where it would be preferable to throw or do a different conversion via extensions.

I've added the strides check.

File: src/lazy.jl [hunk 13]
```diff
@@ -0,0 +1,722 @@
...
+    buf = getbuf(x)
+    len = getlength(buf)
+    opts = getopts(x)
+    b = getbyte(buf, pos)
+    if b != UInt8('{')
+        error = ExpectedOpeningObjectChar
+        @goto invalid
+    end
+    pos += 1
+    @nextbyte
```
[the remark this answers is not in the document]
Comment (quinnj) [reply src/lazy.jl:265]: When I was originally setting this up, it made quite a big difference in the generated LLVM IR to use this consistent pattern of an `@label invalid` with `@goto` vs. lots of separate throw calls. So partly from the maintainability of how optimized we want this code to be, it seems worth it to have it be a macro that "splices" the repetitive code in place to keep clean IR.

File: src/lazy.jl [hunk 14]
```diff
@@ -0,0 +1,722 @@
...
+        la = LazyArray(x)
+        if compact
+            show(io, la)
+        else
+            io = IOContext(io, :compact => true)
+            show(io, MIME"text/plain"(), la)
+        end
+    elseif T == JSONTypes.STRING
+        str, _ = applystring(nothing, x)
+        Base.print(io, "JSON.LazyValue(", repr(convert(String, str)), ")")
```
[the remark this answers is not in the document]
Comment (quinnj) [reply src/lazy.jl:716]: We're relying on the internal rooting/preserving of the `LazyValue` having a reference to `buf`.

File: src/lazy.jl (hunk 3 above)
Comment (quinnj) [reply src/lazy.jl:410]: Performance. It's much slower to use a view in terms of allocating the struct vs. the minimal Ptr + len for the very-short-lived time these are alive. I tried 3-4 different variations of possible ways to handle these and this was the most performant along with the guards to ensure PtrString is never visible at the user-level.

Comment (oscardssmith) [reply src/lazy.jl:410]: I would be curious if a `MemoryRef` based approach could have equal/better perf here.

File: src/lazy.jl (hunk 5 above)
Comment (quinnj) [reply src/lazy.jl:429]: I've opened https://github.com/JuliaLang/julia/issues/58520 to discuss exposing it via `unsafe_symbol`.

File: src/utils.jl (hunk 6 above)
Comment (quinnj) [reply src/utils.jl:44]: Good call. I'll make the compat here more defensive by also checking that `buf` has a `ref` field and that `buf.ref` has a `mem` field. The bigger picture issue here is that it was technically a breaking part of the move to `Memory` that mmapped arrays have the finalizer defined on the `ref.buf` instead of the array itself, so we're trying to deal w/ that breakage here in a way that still makes mmap usage ok for windows users.

File: src/utils.jl [hunk 15]
```diff
@@ -0,0 +1,249 @@
...
+        return false
+    end
+else
+    const _in! = Base.in!
+end
+
+# hand-rolled scoped enum
+module JSONTypes
+    primitive type T 8 end
+    T(x::UInt8) = Base.bitcast(T, x)
```
[the remark this answers is not in the document]
Comment (quinnj) [reply src/utils.jl:17]: Good call; changing to `reinterpret`. It's a force of habit that I always reach for `bitcast` when dealing w/ `primitive type`.

File: src/write.jl (hunk 7 above)
Comment (quinnj) [reply src/write.jl:7]: I've changed `IEEEFloat` to the Union as is defined in Base. We'll have to open another discussion on the `Ryu` usage; it would be great to more formally expose some of that functionality as part of the public API of Base.

File: src/write.jl [hunk 16]
```diff
@@ -0,0 +1,621 @@
...
+    write(io, resize!(buf, pos - 1))
+    return
+end
+
+function json(x; pretty::Union{Integer,Bool}=false, kw...)
+    opts = WriteOptions(; pretty=pretty === true ? 2 : Int(pretty), kw...)
+    _jsonlines_pretty_check(opts.jsonlines, opts.pretty)
+    float_style_check(opts.float_style)
+    y = StructUtils.lower(opts.style, x)
+    buf = Base.StringVector(sizeguess(y))
```
[the remark this answers is not in the document]
Comment (quinnj) [reply src/write.jl:305]: Hmmmm, I tried looking around, but it doesn't seem like there's any easy and _public_ way to currently do this, and eventually you end up at this gnarly definition:
```
@eval _string_n(n::Integer) = $(Expr(:foreigncall, QuoteNode(:jl_alloc_string), Ref{String}, Expr(:call, Expr(:core, :svec), :Csize_t), 1, QuoteNode((:ccall,0x000e)), :(convert(Csize_t, n))))
```
I'll add a defensive check and fallback to Vector{UInt8} if `StringVector` isn't defined.

**Comment by fonsp** [comment]
Awesome work!! I'd like to join the call :)

**Comment by KristofferC** [comment]
A bit confrontative perhaps, but why not a JSON4.jl? To me, looking at the development history, there is nothing really that suggests a JSON5.jl is not in the pipeline (or will that be an update to JSON2.jl?). It seems reasonable to leave this package be quite basic in its functionality that satisfies a lot of people while keeping the JSONX.jl for increasing X being for those that want the most recent cool stuff in JSON parsing.

**Comment by KristofferC** [comment]
@maleadt can we add this repo to nanosoldier so we can run a PkgEval on this?

**Comment by quinnj** [comment]
Aight, finally ironed out the Arrow + JSON extension and added it here + some tests.

Yeah, I'm down for a pkgeval run if that's not too hard. I also should have a decent amount of time over the next week to help with migration efforts. We have the migration guide as part of the docs and I wanted to also start going around and helping to make PRs to update packages as that can uncover a few other things sometimes. I anticipate this will be similar to other large major releases I've helped with in the past for HTTP.jl, CSV.jl, etc.

**Comment by KristofferC** [comment]
I thought with

> providing mostly the same high-level interfaces

it meant that this would be non-breaking with only features on top of it. If this is breaking and requires a major update I see almost no value in doing this change into JSON over another JSON package. Like, who does it benefit? This is literally a whole new package, no?

Regarding

> Let's also be honest about JSON.jl: it's lacked primary maintainership for.......a long time

Surely, the answer to a lack of maintainership isn't to significantly increase the amount of code and functionality?

**Comment by maleadt** [comment]
> @maleadt can we add this repo to nanosoldier so we can run a PkgEval on this?

Sure. Somebody with admin permissions here should reach out to me on Slack.

**Comment by kapple19** [comment]
My 20c opinion: As a simple comparison, HDF5.jl makes sense, JLD2.jl makes sense, JSON3.jl does not.

**Comment by dehann** [comment]
I'm in support of merging this PR.

> it meant that this would be non-breaking with only features on top of it. If this is breaking and requires a major update I see almost no value in doing this change into JSON over another JSON package. Like, who does it benefit? This is literally a whole new package, no?

The original problem was that JSON was split into JSON2 and JSON3.  Maybe there were reasons, but it's also natural that the best code is not available in the beginning.  Yes, there will be some pain to transition back to JSON.jl, but my 2c would be to restore JSON.jl and archive both JSON2 and JSON3.  That way the community can work towards one implementation for one idea "JS Object Notation" -- this is better than trying to maintain 3,4,5.. repos for the same concept.  

> It's 99% non-breaking

Acknowledged that strict semver requires 100% compliance, but in that case again JSON2 and JSON3 should never have happened.

**Comment by tecosaur** [comment]
Given this is a major version change, aren't breaking changes fine, just best minimised without compromising the overall sanity of the new API?

**Comment by ericphanson** [comment]
I'm concerned users like VSCodeServer will struggle with this change because they need to vendor all of their dependencies and this new version has more dependencies. It makes sense to me to have JSON.jl be the "best JSON package for most users" but it seems that JSON.jl is currently "the best JSON package for dependency-sensitive, performance-insensitive users" and such users should be provided a transition plan. Maybe pre-v1 JSON.jl can be spun out into a new SimpleJSON.jl or something?

cc @davidanthoff

File: src/lazy.jl (hunk 14 above)
[the remark this answers is not in the document]
Comment (nhz2) [reply src/lazy.jl:722]: How can I prove that this will not create a use-after-free bug? Is this internal rooting/preserving documented, or is there tooling to ensure this?

**Comment by nhz2** [comment]
I agree that API breakages are fine and expected. Since this is a pre-1.0 package, people depending on this package should expect an unstable API.

My main issue with this PR is that it currently relies on a large number of undocumented features from Julia, which makes reviewing the code almost impossible for someone like me who isn't familiar with how Julia is implemented, or how it will change in the future.

I also don't think mmap should be used in this package. Many commonly used systems do not have mmap available, or working well, and at least on my machine, `read` is much faster for small files. Ref: https://docs.nersc.gov/performance/io/dvs/#do-not-use-memory-mapping-mmap
https://github.com/JuliaLang/julia/issues/58736

```julia-repl
julia> @be _ Mmap.mmap("foo.json") x->GC.gc() seconds=100.0
Benchmark: 710 samples with 1 evaluation
 min    206.200 μs (22 allocs: 1.273 KiB)
 median 268.750 μs (22 allocs: 1.273 KiB)
 mean   298.476 μs (22 allocs: 1.273 KiB)
 max    691.200 μs (22 allocs: 1.273 KiB)

julia> @be _ read("foo.json") x->GC.gc() seconds=100.0
Benchmark: 798 samples with 1 evaluation
 min    158.300 μs (16 allocs: 800 bytes)
 median 203.700 μs (16 allocs: 800 bytes)
 mean   205.402 μs (16 allocs: 800 bytes)
 max    317.600 μs (16 allocs: 800 bytes)

julia> filesize("foo.json")
5

julia> versioninfo()
Julia Version 1.11.6
Commit 9615af0f26 (2025-07-09 12:58 UTC)
Build Info:
  Official https://julialang.org/ release
Platform Info:
  OS: Windows (x86_64-w64-mingw32)
  CPU: 16 × 12th Gen Intel(R) Core(TM) i5-1240P
  WORD_SIZE: 64
  LLVM: libLLVM-16.0.6 (ORCJIT, alderlake)
Threads: 1 default, 0 interactive, 1 GC (on 16 virtual cores)
```

File: Project.toml [hunk 17]
```diff
@@ -1,24 +1,37 @@
 name = "JSON"
 uuid = "682c06a0-de6a-54ab-a142-c8b1cf79cde6"
-version = "0.21.4"
+version = "1.0.0"
 
 [deps]
+Arrow = "69666777-d1a9-59fb-9406-91d4454c9d45"
```
[the remark this answers is not in the document]
Comment (quinnj) [reply Project.toml:6]: Ah, thanks for catching the Arrow inclusion in deps! It's just supposed to be an extra test dependecy.

File: test/regression/issue163.jl [hunk 18]
```diff
@@ -1 +0,0 @@
-@test Float32(JSON.parse(json(2.1f-8))) == 2.1f-8
```
[the remark this answers is not in the document]
Comment (quinnj) [reply test/regression/issue163.jl:1]: I removed a lot of extraneous tests where they were already covered by existing, more consolidated tests (as opposed to one-offs). It seemed like a good time to "re-baseline" some of these issue-specific tests that are now just "standard covered". We also now run both the jsonchecker and JSONTestSuite testing suites, which are industry best-practice for JSON corner cases.

**Comment by quinnj** [comment]
> I'm concerned users like VSCodeServer will struggle with this change because they need to vendor all of their dependencies and this new version has more dependencies. It makes sense to me to have JSON.jl be the "best JSON package for most users" but it seems that JSON.jl is currently "the best JSON package for dependency-sensitive, performance-insensitive users" and such users should be provided a transition plan. Maybe pre-v1 JSON.jl can be spun out into a new SimpleJSON.jl or something?
> 
> cc @davidanthoff

I've added a new `vendor/` directory that includes a single file/module `JSONX` that provides simplified `parse`/`json` functionality with true no-dependency. It's meant to be copied into a project/application and included manually. There are tests specific to the vendor code which will be triggered by a new action workflow that specifically tests the vendored code implementation. cc: @davidanthoff

**Comment by quinnj** [comment]
Usage of and dependency on Mmap has now been removed.

File: src/utils.jl [hunk 19]
```diff
@@ -0,0 +1,246 @@
...
+        Base._setindex!(s.dict, nothing, xT, -idx, sh)
+        return false
+    end
+else
+    const _in! = Base.in!
+end
+
+# hand-rolled scoped enum
+module JSONTypes
+    primitive type T 8 end
```
Comment (nsajko) [inline src/utils.jl:16]: Using `primitive type` instead of just `struct` here seems ugly and unnecessary.

**Comment by nhz2** [comment]
Several packages are currently using the `show_json` interface https://github.com/JuliaIO/JSON.jl/issues/377
It seems like less work to deprecate that interface, instead of completely removing it.

Also, it would be good to include in the migration guide an example of migrating code like https://github.com/JuliaComputing/TableView.jl/blob/a3a3172fed89260b779d79f22c901ed74d96b90d/src/TableView.jl#L293-L319 where `show_json` is used to avoid allocating a temporary dict.

I'm guessing this would involve defining a custom `StructUtils.applyeach` but it wasn't clear to me from https://juliaservices.github.io/StructUtils.jl/dev/#How-make-Works how `applyeach` fits in exactly.
